tests/utils/common/common.py

```diff
@@ -32,6 +28,79 @@
 import traceback
 
 
+class Result:
+    def __init__(self, stdout, stderr, exited):
+        self.stdout = stdout
+        self.stderr = stderr
+        self.exited = exited
+        self.return_code = exited
+
+
+class Connection:
+    def __init__(self, host, user, port, connect_timeout, connect_kwargs={}):
+        self.host = host
+        self.user = user
+        self.port = port
+        self.connect_timeout = connect_timeout
+        self.connect_kwargs = connect_kwargs
+
+        self.key_filename = None
+
+        for k in connect_kwargs.keys():
+            if k == "key_filename":
+                self.key_filename = connect_kwargs[k]
+            else:
+                raise NotImplementedError(f"Argument {k} is not implemented")
+
+    def get_connect_args(self):
+        if self.key_filename is not None:
+            key_arg = ["-i", self.key_filename]
+        else:
+            key_arg = []
+
+        args = (
+            ["ssh"]
+            + key_arg
+            + [
+                "-p",
+                str(self.port),
+                "-o",
+                f"ConnectTimeout={self.connect_timeout}",
+                "-o",
+                "UserKnownHostsFile=/dev/null",
+                "-o",
+                "StrictHostKeyChecking=no",
+                f"{self.user}@{self.host}",
+            ]
+        )
+
+        return args
+
+    def run(self, command, warn=False, hide=False):
+        ssh_command = self.get_connect_args() + [command]
+
+        proc = subprocess.Popen(
+            ssh_command, stdout=subprocess.PIPE, stderr=subprocess.PIPE
+        )
+        stdout, stderr = proc.communicate()
+        stdout, stderr = (stdout.decode(), stderr.decode())
+        if not hide:
+            print(stdout)
+            print(stderr)
+        proc.wait()
+        returncode = proc.returncode
+
+        if returncode == 255:
+            raise ConnectionError(f"Could not connect using command '{ssh_command}'")
+        elif warn is not True and returncode != 0:
```

Comment:
Q: `elif not warn and returncode != 0` ?

Reply:
Maybe also check the returncode first?

Reply:
Yep, that's better, I'll fix it.